### PR TITLE
Diverse flows design

### PR DESCRIPTION
**Pull request by IlyaMoiseev**
Design for #1231

### CODE REVIEW

**Comment by surabujin** [comment]
No coverage for flow reroute - should we allow to "route" diverse flows via same path, or we should allow flow to fail instead?

Should we report client how much flows diverse from our point of view?

And there is no coverage for diverse between 3 and more flows (it is requested in #1231).

**Comment by IlyaMoiseev** [comment]
Actually, reroute algorithm performs as usual for diverse flows, all changes are in Available network itself.

Yes, we return `diverse path computation strategy` to client (with diverse level: switch, isl or partial). 
But for reroute case this information will be omitted, and diverse level may me decreased. @pzakatov  Should we allow that behavior, or keep flow in fail state?

It's not depends from flows count. All diverse flows should have links to other flows in the diverse group. This flows are fetched and excluded from Available network.

File: docs/design/solutions/pce-diverse-flows/pce-diverse-flows.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Ability to create diverse flows
+
+## Goals
+Have redundancy on flows: different flows reside on different switches and/or ISLs
+Main goal is to find shortest diverse path
+
+## NB contract changes
+Request: add flow id to make new flow diverse with
```
Comment (sergii-iakovenko) [inline docs/design/solutions/pce-diverse-flows/pce-diverse-flows.md]: Shouldn't it be optional?

File: docs/design/solutions/pce-diverse-flows/pce-diverse-flows-create.puml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    Computer -> DB: get active links
+    DB --> Computer
+    create Network
+    Computer -> Network: create from retrieved links
+    Network --> Computer
+    Computer --> Storm: AvailableNetwork
+    deactivate Computer
+end
+
+Storm -> DB: get diverse with flow by id
```
Comment (sergii-iakovenko) [inline docs/design/solutions/pce-diverse-flows/pce-diverse-flows-create.puml]: Could you please rephrase it? Get what?

File: docs/design/solutions/pce-diverse-flows/pce-diverse-flows-create.puml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+alt diverse with flow has flow group
+    Storm -> DB: get flows with the same flow group
+    DB --> Storm: diverse with flows
```
Comment (sergii-iakovenko) [inline docs/design/solutions/pce-diverse-flows/pce-diverse-flows-create.puml]: Could you please rephrase it? Diverse what?

File: docs/design/solutions/pce-diverse-flows/pce-diverse-flows-create.puml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+group Find path
+    Storm -> Computer: getPathWithAvailableNetwork
+    activate Computer
+    Computer -> Computer: PathFinder.findPathInNetwork
+    Computer --> Storm: computed PathPair
+    deactivate Computer
+end
+
+Storm -> IntersectionComputer: get intersection counts by ISLs and switches in group
```
Comment (sergii-iakovenko) [inline docs/design/solutions/pce-diverse-flows/pce-diverse-flows-create.puml]: Does it work with AvailableNetwork? If not, please indicate how it detects the intersections.

**Comment by IlyaMoiseev** [comment]
Updated. Diverse processing now uses Flow Segments as input.
Comments fixed

File: docs/design/solutions/pce-diverse-flows/pce-diverse-flows.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+Response: applied diverse path computation strategy for computed diverse path
```
[the remark this answers is not in the document]
Comment (pzakatov) [reply docs/design/solutions/pce-diverse-flows/pce-diverse-flows.md]: yep. looks better

[the remark this answers is not in the document]
Comment (pzakatov) [reply docs/design/solutions/pce-diverse-flows/pce-diverse-flows.md]: please fix

**Comment by carmine** [comment]
A couple of questions:

1. Is this discussion reflected in the design doc for diverse paths?
2. I haven't fully digested all the comments above (hard to follow), but the question is whether PCE should reflect similar design principles used in schedulers, e.g. for openstack or kubernetes. Maybe the design is already there .. will review.
